File: lab6/c/pq.c
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <string.h>
#include <stdarg.h>
#include <stdbool.h>
#include <assert.h>

#include "general.h"
#include "graph.h"
#include "pq.h"


typedef struct {size_t i;} hidx_t;

size_t idx_of(hidx_t i) {return i.i - 1;}

const hidx_t INVALID_HIDX = {0};
bool hidx_is_invalid(hidx_t i) {return i.i == 0;}

const hidx_t hidx_first = {1};



// 1-based operations
hidx_t parent(hidx_t i) {return (hidx_t){i.i / 2}; }
hidx_t left(hidx_t i) {return (hidx_t){i.i * 2}; }
hidx_t right(hidx_t i) {return (hidx_t){i.i * 2 + 1}; }
bool has_parent(hidx_t i) {return parent(i).i>0; }
bool has_left(size_t n, hidx_t i) {return left(i).i<=n; }
bool has_right(size_t n, hidx_t i) {return right(i).i<=n; }


struct _DPQ_t {
    size_t num_elem;     // Number of elements
    size_t heap_size;    // Current size of heap
    weight_t *D;  // Priorities
    node_t *H;    // The heap
    hidx_t *I;    // Index of element in heap.
};


void DPQ_init(DPQ_t *pq, size_t num_elem) {
    pq->num_elem = num_elem;
    pq->heap_size=0;
    pq->D = CALLOC(weight_t,num_elem);
    pq->H = CALLOC(node_t,num_elem);
    pq->I = CALLOC(hidx_t,num_elem);

    for (size_t i=0;i<num_elem;++i) {
      pq->I[i]=INVALID_HIDX;
      pq->D[i]=weight_inf();
    }

}

void DPQ_uninit(DPQ_t *pq) {
    free(pq->D);
    free(pq->H);
    free(pq->I);
}

DPQ_t *DPQ_new(size_t num_elem) {
    DPQ_t *pq = CALLOC(DPQ_t,1);
    DPQ_init(pq,num_elem);
    return pq;
}

void DPQ_delete(DPQ_t *pq) {
    DPQ_uninit(pq);
    free(pq);
}

weight_t *DPQ_dist_free(DPQ_t *pq) {
  weight_t *res=pq->D;
  free(pq->H);
  free(pq->I);
  free(pq);
  return res;
}


weight_t DPQ_prio(DPQ_t const *pq, node_t u) {
    assert(u<pq->num_elem);
    return pq->D[u];
}

weight_t _DPQ_hprio(DPQ_t const *pq, hidx_t i) {
    return DPQ_prio(pq,pq->H[idx_of(i)]);
}

void _DPQ_adjustI(DPQ_t *pq, hidx_t i) {
    pq->I[pq->H[idx_of(i)]] = i;
}

void _DPQ_swap(DPQ_t *pq, hidx_t i, hidx_t j) {
    node_t t = pq->H[idx_of(i)];
    pq->H[idx_of(i)] = pq->H[idx_of(j)];
    pq->H[idx_of(j)] = t;
    _DPQ_adjustI(pq,i);
    _DPQ_adjustI(pq,j);
}


void _DPQ_sift_up(DPQ_t *pq, hidx_t i) {
    while (has_parent(i) && weight_less(_DPQ_hprio(pq,i), _DPQ_hprio(pq,parent(i)))) {
        _DPQ_swap(pq,i,parent(i));
        i = parent(i);
    }
}

void _DPQ_sift_down(DPQ_t *pq, hidx_t i) {
    while (has_right(pq->heap_size,i)) {
        weight_t w = _DPQ_hprio(pq,i);
        weight_t lw = _DPQ_hprio(pq,left(i));
        weight_t rw = _DPQ_hprio(pq,right(i));

        if (!weight_less(lw, w) && !weight_less(rw, w)) return;

        if (weight_less(lw,rw)) {
            _DPQ_swap(pq,i,left(i));
            i=left(i);
        } else {
            _DPQ_swap(pq,i,right(i));
            i=right(i);
        }
    }

    if (has_left(pq->heap_size,i)) {
        weight_t w = _DPQ_hprio(pq,i);
        weight_t lw = _DPQ_hprio(pq,left(i));
        if (weight_less(lw,w)) _DPQ_swap(pq,i,left(i));
    }

}

bool DPQ_contains(DPQ_t const *pq, node_t u) {
    assert(u<pq->num_elem);
    return !hidx_is_invalid(pq->I[u]);
}

hidx_t _DPQ_last_idx(DPQ_t const *pq) {
    return (hidx_t){pq->heap_size};
}

void DPQ_insert(DPQ_t *pq, node_t u, weight_t w) {
    assert(u<pq->num_elem);
    assert(!DPQ_contains(pq,u));
    pq->D[u]=w;
    ++pq->heap_size;
    hidx_t i = _DPQ_last_idx(pq);
    pq->H[idx_of(i)]=u;
    _DPQ_adjustI(pq,i);
    _DPQ_sift_up(pq,i);
}

bool DPQ_is_empty(DPQ_t const *pq) {return pq->heap_size==0;}

node_t DPQ_pop_min(DPQ_t *pq) {
    assert(!DPQ_is_empty(pq));
    hidx_t li = _DPQ_last_idx(pq);
    _DPQ_swap(pq,hidx_first,li);
    node_t res = pq->H[idx_of(li)];
    --pq->heap_size;
    pq->I[res] = INVALID_HIDX;
    _DPQ_sift_down(pq,hidx_first);
    return res;
}

void DPQ_decrease_key(DPQ_t *pq, node_t u, weight_t w) {
    assert(u<pq->num_elem);
    assert(DPQ_contains(pq,u));
    assert(weight_less(w, pq->D[u]));
    pq->D[u]=w;

    hidx_t i = pq->I[u];
    _DPQ_sift_up(pq,i);

}

File: lab6/python/pq.py
import sys

import general
import weight

class hidx_t:
    def __init__(self,i):
        self.i = i

def idx_of(i):
    return i.i - 1
    
INVALID_HIDX = hidx_t(0)
def hidx_is_invalid(i):
    return i.i == 0
    
hidx_first = hidx_t(1)

# 1-based operations
def parent(i):
    return hidx_t(i.i //2)
def left(i):
    return hidx_t(2*i.i)
def right(i):
    return hidx_t((2*i.i) + 1)
def has_parent(i):
    return parent(i).i > 0
def has_left(n, i):
    return left(i).i<= n
def has_right(n, i):
    return right(i).i <= n

class DPQ_t:
    ##
     # Create new, initially empty priority queue.
     # The priorities of all nodes < N are set to infinite.
     # @param N Maximum number of nodes
     #/
    def __init__(self, size):
        self.num_elem = size
        self.heap_size = 0
        self.D = [None]*size
        self.H = [None]*size
        self.I = [None]*size
        
        for i in range(0, self.num_elem):
            self.I[i] = INVALID_HIDX
            self.D[i] = weight.weight_inf()
        
    def DPQ_dist_free(self):
        res = self.D
        return res

    ##
     # Get priority of node.
     #
     # Complexity: O(1)
     #
     # @pre u < N
     #/
    def DPQ_prio(self, u):
        assert(u<self.num_elem)
        return self.D[u]
        
    def _DPQ_hprio(self, i):
        return self.DPQ_prio(self.H[idx_of(i)])

    def _DPQ_adjustI(self, i):
        self.I[self.H[idx_of(i)]] = i
        
    # Helper function for swapping nodes in the tree
    def _DPQ_swap(self, i, j):
        t = self.H[idx_of(i)]
        self.H[idx_of(i)] = self.H[idx_of(j)]
        self.H[idx_of(j)] = t
        
        self._DPQ_adjustI(i)
        self._DPQ_adjustI(j)
        
    def _DPQ_sift_up(self, i):
        while (has_parent(i) and weight.weight_less(self._DPQ_hprio(i), self._DPQ_hprio(parent(i)))):
            self._DPQ_swap(i, parent(i))
            i = parent(i)
        
    def _DPQ_sift_down(self, i):
        while (has_right(self.heap_size, i)):
            w = self._DPQ_hprio(i)
            lw = self._DPQ_hprio(left(i))
            rw = self._DPQ_hprio(right(i))
            if (not weight.weight_less(lw, w) and not weight.weight_less(rw, w)):
                # No children, we're finisehd
                return
                
            if (weight.weight_less(lw, rw)):
                self._DPQ_swap(i, left(i))
                i = left(i)
            else:
                self._DPQ_swap(i, right(i))
                i = right(i)
                
        if (has_left(self.heap_size,i)):
            w = self._DPQ_hprio(i)
            lw = self._DPQ_hprio(left(i))
            if (weight.weight_less(lw, w)):
                self._DPQ_swap(i, left(i))

    ##
     # Check if node is contained in queue
     #
     # Complexity: O(1)
     #
     # @pre u < N
     #/
    def DPQ_contains(self, u):
        assert(u<self.num_elem)
        # Linear Search required as unordered
        return not hidx_is_invalid(self.I[u])
        
    def _DPQ_last_idx(self):
        return hidx_t(self.heap_size)

    ##
     # Insert node into queue, with given priority.
     # The priority of this node is set accordingly.
     #
     # Complexity: O(log N)
     #
     # @pre Node must not be in queue
     # @pre u < N
     # @param u node
     # @param w priority
     #/
    def DPQ_insert(self, u, w):
        assert(u < self.num_elem)
        assert(not self.DPQ_contains(u))
        self.D[u]=w
        self.heap_size = self.heap_size + 1
        i = self._DPQ_last_idx()
        self.H[idx_of(i)] = u
        self._DPQ_adjustI(i)
        self._DPQ_sift_up(i)

    #
     # Check if queue is empty
     #
     # Complexity: O(1)
     #/
    def DPQ_is_empty(self):
        return self.heap_size == 0

    ##
     # Retrieve and remove an element of the queue with minimal priority
     #
     # Complexity: O(log N)
     #
     # @pre The queue is not empty
     #/
    def DPQ_pop_min(self):
        assert(not self.DPQ_is_empty())
        li = self._DPQ_last_idx()
        self._DPQ_swap(hidx_first, li)
        res = self.H[idx_of(li)]
        self.heap_size = self.heap_size - 1
        self.I[res] = INVALID_HIDX
        self._DPQ_sift_down(hidx_first)
        return res

    ##
     # Lower priority of element in queue
     #
     # Complexity: O(log N)
     #
     # @pre the node is in the queue
     # @pre the new priority is less than the current nodes priority
     #/
    def DPQ_decrease_key(self, u, w):
        assert(u < self.num_elem)
        assert(self.DPQ_contains(u))
        assert(weight.weight_less(w, self.D[u]))
        self.D[u] = w
        
        i = self.I[u]
        self._DPQ_sift_up(i)
    
